app_dart/lib/src/service/scheduler.dart

```diff
@@ -196,6 +198,60 @@ class Scheduler {
     return SchedulerConfig.fromBuffer(configBytes);
   }
 
+  /// Get all postsubmit targets that should be immediately started for [Commit].
+  List<Target> _getInitialPostSubmitTargets(Commit commit, SchedulerConfig config) {
+    final List<Target> postsubmitTargets = getPostSubmitTargets(commit, config);
+
+    // Filter targets to only those without dependencies.
+    final List<Target> initialTargets =
+        postsubmitTargets.where((Target target) => target.dependencies.isEmpty).toList();
+    return initialTargets;
+  }
+
+  /// Get all targets run on postsubmit for [Commit].
+  ///
+  /// Filter [SchedulerConfig] to only the targets expected to run for the branch,
+  /// and that do not have any dependencies.
+  List<Target> getPostSubmitTargets(Commit commit, SchedulerConfig config) {
+    // Filter targets to only those run in postsubmit.
+    final List<Target> postsubmitTargets = config.targets.where((Target target) => target.postsubmit).toList();
+    return _filterEnabledTargets(commit, config, postsubmitTargets);
+  }
+
+  /// Get all targets run on presubmit for [Commit].
+  ///
+  /// Filter [SchedulerConfig] to only the targets expected to run for the branch,
+  /// and that do not have any dependencies.
+  List<Target> getPreSubmitTargets(Commit commit, SchedulerConfig config) {
+    // Filter targets to only those run in presubmit.
+    final Iterable<Target> presubmitTargets = config.targets.where((Target target) => target.presubmit);
+
+    return _filterEnabledTargets(commit, config, presubmitTargets.toList());
+  }
+
+  /// Filter [targets] to only those that are expected to run for [Commit] with [SchedulerConfig].
+  ///
+  /// A [Target] is expected to run if:
+  ///   1. [Target.enabledBranches] exists and matches [Commit].
+  ///   2. Otherwise, [SchedulerConfig.enabledBranches] matches [Commit].
+  List<Target> _filterEnabledTargets(Commit commit, SchedulerConfig config, List<Target> targets) {
+    final List<Target> filteredTargets = <Target>[];
+
+    // 1. Add targets with local definition
+    final Iterable<Target> overrideBranchTargets = targets.where((Target target) => target.enabledBranches.isNotEmpty);
+    final Iterable<Target> enabledTargets =
+        overrideBranchTargets.where((Target target) => target.enabledBranches.contains(commit.branch));
+    filteredTargets.addAll(enabledTargets);
+
+    // 2. Add targets with global definition
+    if (config.enabledBranches.contains(commit.branch)) {
+      final Iterable<Target> defaultBranchTargets = targets.where((Target target) => target.enabledBranches.isEmpty);
```

Comment:
Ah, I see, it's because you're first filtering for only targets that don't locally define it.